src/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.OSX.cs

```diff
@@ -174,16 +172,106 @@ private sealed class RunningInstance
-                        Interop.RunLoop.CFRunLoopStop(_watcherRunLoop);
+                        s_scheduledStreamsCount = 1;
+                        var runLoopStarted = new ManualResetEventSlim();
+                        new Thread(WatchForFileSystemEventsThreadStart) { IsBackground = true }.Start(new object[] { runLoopStarted, eventStream });
+                        runLoopStarted.Wait();
+                    }
+                }
+
+                public static void UnscheduleFromRunLoop(SafeEventStreamHandle eventStream)
+                {
+                    if (s_watcherRunLoop != IntPtr.Zero)
+                    { 
+                        lock (s_lockObject)
+                        {
+                            if (s_watcherRunLoop != IntPtr.Zero)
+                            { 
+                                // Always unschedule the RunLoop before cleaning up
+                                Interop.EventStream.FSEventStreamUnscheduleFromRunLoop(eventStream, s_watcherRunLoop, Interop.RunLoop.kCFRunLoopDefaultMode);
+                                s_scheduledStreamsCount--;
+
+                                if (s_scheduledStreamsCount == 0)
+                                {
+                                    // Stop the FS event message pump
+                                    Interop.RunLoop.CFRunLoopStop(s_watcherRunLoop);                                    
+                                    s_watcherRunLoop = IntPtr.Zero;
+                                }
+                            }
+                        }
+                    }
+                }
+
+                private static void WatchForFileSystemEventsThreadStart(object args)
+                {
+                    var inputArgs = (object[])args;
+                    var runLoopStarted = (ManualResetEventSlim)inputArgs[0];
+                    var _eventStream = (SafeEventStreamHandle)inputArgs[1];
+                    // Get this thread's RunLoop
+                    var runLoop = Interop.RunLoop.CFRunLoopGetCurrent();
+                    s_watcherRunLoop = runLoop;
+                    Debug.Assert(s_watcherRunLoop != IntPtr.Zero);
+
+                    // Retain the RunLoop so that it doesn't get moved or cleaned up before we're done with it.
+                    IntPtr retainResult = Interop.CoreFoundation.CFRetain(runLoop);
+                    Debug.Assert(retainResult == runLoop, "CFRetain is supposed to return the input value");
+
+
+                    // Schedule the EventStream to run on the thread's RunLoop
+                    Interop.EventStream.FSEventStreamScheduleWithRunLoop(_eventStream, runLoop, Interop.RunLoop.kCFRunLoopDefaultMode);
+
+                    runLoopStarted.Set();
+                    try
+                    {
+                        // Start the OS X RunLoop (a blocking call) that will pump file system changes into the callback function
+                        Interop.RunLoop.CFRunLoopRun();
```

Comment:
Does this actually throw?  If so, right now the app will crash when it does.  Are we ok with that? ("yes" might be an ok answer.)

Reply:
Previously it was within try-finally section and I just left it as is. But in other place like [NetworkAddressChange.OSX.cs](https://github.com/dotnet/corefx/blob/fc1574ecf3d0b9a9035bc04abae465f27c510dab/src/System.Net.NetworkInformation/src/System/Net/NetworkInformation/NetworkAddressChange.OSX.cs#L199) `Interop.RunLoop.CFRunLoopRun` is invoked without error handling.